subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy

```diff
@@ -0,0 +1,166 @@
+    doLast {
+        outputFile.parentFile.mkdirs()
+        outputFile.text = "produced"
+    }
+}
+
+task consumer {
+    def inputFile = file("${consumedLocation}")
+    def outputFile = file("consumerOutput.txt")
+    inputs.files(inputFile)
+    outputs.file(outputFile)
+    doLast {
+        outputFile.text = "consumed"
+    }
+}
+        """
+
+        when:
+        executer.expectDocumentedDeprecationWarning(DEPRECATION_WARNING)
+        then:
+        succeeds("producer", "consumer")
+
+        when:
+        executer.expectDocumentedDeprecationWarning(DEPRECATION_WARNING)
+        then:
+        succeeds("consumer", "producer")
+
+        where:
+        description            | producerOutput     | producedLocation           | consumedLocation
+        "same location"        | "file(outputFile)" | "output.txt"               | "output.txt"
+        "consuming ancestor"   | "file(outputFile)" | "build/dir/sub/output.txt" | "build/dir"
+        "consuming descendant" | "dir('build/dir')" | "build/dir/sub/output.txt" | "build/dir/sub/output.txt"
+    }
+
+    def "does not detect missing dependency when consuming the sibling of the output of the producer"() {
+        buildFile << """
+            task producer {
+                def outputFile = file("build/output.txt")
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.parentFile.mkdirs()
+                    outputFile.text = "produced"
+                }
+            }
+
+            task consumer {
+                def inputFile = file("build/notOutput.txt")
+                def outputFile = file("consumerOutput.txt")
+                inputs.files(inputFile)
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.text = "consumed"
+                }
+            }
+        """
+
+        expect:
+        succeeds("producer", "consumer")
+        succeeds("consumer", "producer")
+    }
+
+    def "dependency chain counts as dependency (including #dependency)"() {
+        buildFile << """
+            task producer {
+                def outputFile = file("output.txt")
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.text = "produced"
+                }
+            }
+
+            task consumer {
+                def inputFile = file("output.txt")
+                def outputFile = file("consumerOutput.txt")
+                inputs.files(inputFile)
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.text = "consumed"
+                }
+            }
+
+            task a
+            task b
+            task c
+            task d
+
+            consumer.dependsOn(d)
+
+            d.dependsOn(c)
+            ${dependency}
+            b.dependsOn(a)
+
+            a.dependsOn(producer)
+        """
+
+        expect:
+        succeeds("producer", "b", "c", "consumer")
```

Comment:
Why don't we just invoke `consumer`?

Reply:
Because for `finalizedBy` and `mustRunAfter`, the intermediate tasks are not triggered.

Reply:
I added a comment.

Reply:
Thanks. Another way to disambiguate would be if the list of tasks to execute would be a parameter to the test.